### PR TITLE
Add removal of unmapped groups

### PR DESCRIPTION
**Pull request by bosvos**

### CODE REVIEW

File: msaad.go [hunk 1]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
 	groupsChanged := false
 	userHasAnyIMQSPermission := false
 
+	// identify unmapped groups
+	removeIDs, _ := DecodePermit(make([]byte, 0))
+	allowedIDs, _ := DecodePermit(make([]byte, 0))
+
+	m.parent.Log.Debugf("MSAAD empty role arrays constructed")
```
Comment (jasonkofo) [inline msaad.go:532]:
```suggestion
	if m.parent.MSAAD.Config.Verbose {
    	    m.parent.Log.Debugf("MSAAD empty role arrays constructed")
    	}
```

File: msaad.go [hunk 2]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+
+	// get all mapped group ids
+	for _, internalGroupName := range m.Config.RoleToGroup {
+		m.parent.Log.Debugf("MSAAD checking all roles: %v", internalGroupName)
```
Comment (jasonkofo) [inline msaad.go:536]: I think this class of logs should not be allowed to bloat the configs, unless we explicitly want them. So I will just let you know all of the logs that I am sure should be masked behind verbose config.
```suggestion
		if m.parent.MSAAD.Config.Verbose {
		    m.parent.Log.Debugf("MSAAD checking all roles: %v", internalGroupName)
		}
```

File: msaad.go [hunk 3]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+		if !ok {
+			// We've already logged an error about this, so here we just ignore it
+			continue
+		}
+		m.parent.Log.Debugf("MSAAD add allowed ID for %v", internalGroupName)
+		allowedIDs = append(allowedIDs, internalGroup.ID)
+	}
+
+	for _, groupName := range m.Config.DefaultRoles {
+		m.parent.Log.Debugf("MSAAD checking default roles: %v", groupName)
```
Comment (jasonkofo) [inline msaad.go:548]: Mask behind verbose config

File: msaad.go [hunk 4]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+
+		internalGroup, ok := roleGroups.nameToGroup[groupName]
+		if !ok {
+			// We've already logged an error about this, so here we just ignore it
+			continue
+		}
+		m.parent.Log.Debugf("MSAAD add allowed default ID for %v", groupName)
```
Comment (jasonkofo) [inline msaad.go:555]: Mask behind verbose config

File: msaad.go [hunk 5]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+		allowedIDs = append(allowedIDs, internalGroup.ID)
+	}
+
+	// now remove all IDs from groupID that is NOT in allowedIDs
+	for _, groupID := range groupIDs {
+		if allowedIDs.IndexOf(groupID) == -1 {
+			m.parent.Log.Debugf("MSAAD unmapped ID %v, add to remove list", groupID)
```
Comment (jasonkofo) [inline msaad.go:562]: Mask behind verbose config

File: msaad.go [hunk 6]
```diff
@@ -523,6 +525,56 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+			removeIDs = append(removeIDs, groupID)
+		}
+	}
+
+	for _, id := range removeIDs {
+		if idx := groupIDs.IndexOf(id); idx != -1 {
+			group := roleGroups.idToGroup[id]
+			groupName := group.Name
```
Comment (jasonkofo) [inline msaad.go:570]: Superfluous creation of the variable here. Rather just pass the ltype in the next line.

**Review by jasonkofo — CHANGES_REQUESTED** [review]

**Comment by bosvos** [comment]
Changes made - please review.

**Comment by bosvos** [comment]
@jasonkofo I suppose we need to get this into master.

File: msaad.go [hunk 7]
```diff
@@ -523,6 +524,64 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+			m.parent.Log.Debugf("MSAAD checking all roles: %v", internalGroupName)
+		}
+
+		internalGroup, ok := roleGroups.nameToGroup[internalGroupName]
+		if !ok {
+			// We've already logged an error about this, so here we just ignore it
+			continue
+		}
+		m.parent.Log.Debugf("MSAAD add allowed ID for %v", internalGroupName)
+		allowedIDs = append(allowedIDs, internalGroup.ID)
```
Comment (jasonkofo) [inline msaad.go:547]: Please excuse my being pedantic about issues of style
```suggestion
		if internalGroup, ok := roleGroups.nameToGroup[internalGroupName]; ok {
		    m.parent.Log.Debugf("MSAAD add allowed ID for %v", internalGroupName)
		    allowedIDs = append(allowedIDs, internalGroup.ID)
		}
```

File: msaad.go [hunk 8]
```diff
@@ -503,6 +503,7 @@ func removeFromGroupList(list []GroupIDU32, i int) []GroupIDU32 {
 
 func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, internalUserID UserId) error {
 	nameInLogs := aadUser.profile.bestEmail()
+	m.parent.Log.Debugf("MSAAD syncRoles started for %s", nameInLogs)
```
Comment (jasonkofo) [inline msaad.go:506]: A part of me thinks that including this at a debug level will introduce LOADS of noise into the logs, especially if one is expecting 120something of these records every 3 minutes

I wanna recommend that we drop the log level of this one. 
```suggestion
	m.parent.Log.Tracef("MSAAD syncRoles started for %s", nameInLogs)
```

File: msaad.go [hunk 9]
```diff
@@ -523,6 +524,64 @@ func (m *MSAAD) syncRoles(roleGroups *cachedRoleGroups, aadUser *msaadUser, inte
...
+		if m.parent.MSAAD.Config.Verbose {
+			m.parent.Log.Debugf("MSAAD checking all roles: %v", internalGroupName)
+		}
+
+		internalGroup, ok := roleGroups.nameToGroup[internalGroupName]
+		if !ok {
+			// We've already logged an error about this, so here we just ignore it
+			continue
+		}
+		m.parent.Log.Debugf("MSAAD add allowed ID for %v", internalGroupName)
```
Comment (jasonkofo) [inline msaad.go:546]:
```suggestion
		m.parent.Log.Tracef("MSAAD add allowed ID for %v", internalGroupName)
```

**Review by jasonkofo — CHANGES_REQUESTED** [review]
Seriously minor issues that I have

**Review by jasonkofo — APPROVED** [review]

**Pull request merged**